Add library codegen docs for Kotlin, Swift, TypeScript (#1698)
avohq/docs: pages/implementation/library-codegen.mdx

# Library codegen for Kotlin, Swift, and TypeScript

Avo can emit generated code in two shapes for Kotlin, Swift, and TypeScript sources: the default **single-source** output (one file per source) and the **library mode** output that splits the runtime from the per-source types. Both modes are supported — they serve different use cases and you can generate either from the same source.

Library mode is the better fit when you have multiple Avo sources sharing the same app, when you're building an internal tracking library reused across apps, or when you want the runtime code separated from the per-source event types.

---

## 1. The shape of library mode

> `avo pull` emits two files per source (three for TypeScript) instead of one. The shared source-independent runtime — `Avo` class, validation helpers, `AvoEvent` protocol/interface, `AvoInvoke` — lives in a separate "library" file. The per-source "app" file contains only the things that change when your tracking plan changes: per-event classes and the codegen-bound IDs.

Reviewer: ```suggestion
`avo pull` emits multiple files per source instead of one. The shared source-independent runtime — `Avo` class, validation helpers, `AvoEvent` protocol/interface, `AvoInvoke` — lives in a separate "library" file. The per-source "app" file contains only the things that change when your tracking plan changes: per-event classes and the codegen-bound IDs. Optionally, third file is extracted, with everything needed for Avo initialisation, so initialisation can also be decoupled from the apps.
```

Author: Applied in 54db72a5. Used your wording verbatim — "emits multiple files per source" and the optional-third-file framing.

### What gets generated

> - **Library file**: the runtime. `AvoEnv`, `AvoAssert`, `AvoInvoke`, the `AvoEvent` protocol/interface, and the `Avo` class. Regenerated rarely. Shared across all sources/apps. Optionally orchestrated from a single place — "the library".

Reviewer: ```suggestion
- **Library file**: the runtime. `AvoEnv`, `AvoAssert`, `AvoInvoke`, the `AvoEvent` protocol/interface, and the `Avo` class. Regenerated rarely. Every source will generate identical file. Shared across all sources/apps. Optionally orchestrated from a single place — "the library".
```

Author: Applied in 54db72a5. Added "Every source generates an identical library file." Slight phrasing tweak from your suggestion but same meaning.

import codegenConfig from './AvoConfig';

const avo = Avo.init(
  {
    env: AvoEnv.Dev,
    destinations: {
      Custom: {
        make(env, apiKey) { /* initialize your SDK */ },
        logEvent(event) { myAnalyticsSDK.logEvent(event.name, event.properties); },
        setUserProperties(userId, props) { myAnalyticsSDK.identify(userId, props); },
      },
    },
  },
  codegenConfig,
);

try {
  avo.track(new LoginSuccessEvent(1730000000, 'team_42', 'acme.example'));
} catch (e) {
  if (e instanceof AvoVerificationError) {
    console.error('[avo]', e.messages.join(', '));
  } else {
    throw e;
  }
}
```

`AvoDestination` (`make`, `logEvent`, `setUserProperties`) plays the same role as the custom destination interface in the single-source output — the shape is different but the idea is the same. `destinations` is all-or-nothing: if you provide any, you must provide an implementation for every `DestinationKey` or `Avo.init` will throw.

If the all-or-nothing model doesn't fit (for example, you want to fan out manually for some destinations), omit `destinations` and fan out yourself:

```ts
const avo = Avo.init({ env: AvoEnv.Dev }, codegenConfig);

> const payloads = avo.track(new LoginSuccessEvent(1730000000, 'team_42', 'acme.example'));

Reviewer: Wrap this one in try catch too for better reference

Author: Fixed in 54db72a5. Wrapped the no-destinations TypeScript example in try/catch — matches the destinations-wired example above it.

In Swift and Kotlin the returned map is the *only* thing the runtime gives you — fan-out to your analytics SDKs is your job. In TypeScript you can either fan out yourself or pass `AvoDestination` implementations into `Avo.init()` and let the runtime call them.

### When it isn't the right fit

- You're on a language other than Kotlin, Swift, or TypeScript. The library approach is implemented only for those three.
- You rely on the previous shape of the `Avo` class (e.g. `avo.loginSuccess(...)` method calls). You'll need to update call sites — see the migration section below.
- You want the runtime and per-event code in a single file for build-system reasons. The legacy single-file codegen is still supported; don't opt in.

---

## 2. Quick start

Pull library-shaped code for an existing source by enabling the library-interface mode on that source, then run `avo pull`. The CLI will emit the library file alongside the existing source file (and `AvoConfig.ts` for TypeScript).

> **Note:** the exact opt-in flag name surfaced to customers isn't covered here — see Questions / gaps at the end.

> 

Author: Partial fix in 6baeb4cc. The doc is a draft — the customer-facing opt-in flag and Swift/Kotlin module-packaging details are deliberately deferred to the [Questions / gaps](https://github.com/avohq/docs/blob/worktree-library-codegen-approach/pages/implementation/library-codegen.mdx#questions--gaps) section at the bottom (items 1–3, 7). Those need answers from someone who knows the CLI rollout plan, which I do not. Added a "Draft — internal review" callout at the top of the page so anyone landing on it sees the same warning the reviewer raised: do not link this from customer-facing places until the gaps are filled.

    timestamp = 1730000000,
    teamId = "team_42",
    teamDomain = "acme.example",
))
payloads["custom"]?.let { event ->
    myAnalyticsSdk.logEvent(event.name, event.properties)
}
```

#### Swift

```swift
// Before
let avo = Avo(env: .dev, customDestination: MyCustomDestination())
avo.loginSuccess(
    timestamp: 1730000000,
    teamId: "team_42",
    teamDomain: "acme.example"
)

// After
let avo = Avo.initAvo(env: .dev)
let payloads = try avo.process(event: LoginSuccessEvent(
    timestamp: 1730000000,
    teamId: "team_42",
    teamDomain: "acme.example"
))
if let event = payloads["custom"] {
    MyAnalyticsSDK.logEvent(name: event.name, properties: event.properties)
}
```

#### TypeScript

```ts
// Before
import Avo from './Avo';
Avo.initAvo({ env: Avo.AvoEnv.Dev }, /* destinations */ {});
Avo.loginSuccess({
  timestamp: 1730000000,
  teamId: 'team_42',
  teamDomain: 'acme.example',
});

// After
import { Avo, AvoEnv, LoginSuccessEvent } from './Avo';
import codegenConfig from './AvoConfig';

const avo = Avo.init({ env: AvoEnv.Dev }, codegenConfig);
avo.track(new LoginSuccessEvent(1730000000, 'team_42', 'acme.example'));
```

### Checklist of changes in your app

> 1. **Drop the legacy `Avo` constructor**. Replace direct `Avo(...)` / `Avo.initAvo(...)` calls with the new init that takes no destinations (Swift, Kotlin) or accepts the codegen-bound config from `AvoConfig.ts` (TypeScript).

Author: Fixed in 6baeb4cc. Reworded the checklist item — only the legacy `Avo(env:, customDestination:, ...)` constructor (Swift/Kotlin) and the legacy free-function `Avo.initAvo(config, destinations)` (TypeScript) are deprecated. `Avo.initAvo(...)` stays as the target for Kotlin/Swift (the generated extension on the library `Avo` class); TypeScript switches to `Avo.init(config, codegenConfig)`.

# Library codegen for Kotlin, Swift, and TypeScript

> **Draft — internal review.** This page describes the library-shaped codegen output but does not yet include the customer-facing opt-in flag, the recommended Swift module / Kotlin Gradle layout, or the destination-wrapper details. Those open items are listed in [Questions / gaps](#questions--gaps) at the bottom. Don't link this page from customer-facing places until it's filled in.

> This page covers the library approach for generated tracking code. It's a different shape of output for Kotlin, Swift, and TypeScript sources. The previous (single-file) codegen is still supported.

Reviewer: We can probably call the old one - single-sorce, and the new one multi-source.

Author: Addressed in a27de02d. Reframed the doc: single-source (default) vs library mode (alternative for multi-source / shared library setups). Both supported; library is no longer positioned as a successor.

---

## 1. Why the library approach?

> `avo pull` now emits two files per source (three for TypeScript) instead of one. The shared runtime — `Avo` class, validation helpers, `AvoEvent` protocol/interface, `AvoInvoke` — lives in a separate "library" file. The per-source "app" file contains only the things that change when your tracking plan changes: per-event classes, the codegen-bound IDs, and (for TypeScript) the `DestinationKey` union.

Reviewer: ```suggestion
`avo pull` now emits two files per source (three for TypeScript) instead of one. The shared source-independent runtime — `Avo` class, validation helpers, `AvoEvent` protocol/interface, `AvoInvoke` — lives in a separate "library" file. The per-source "app" file contains only the things that change when your tracking plan changes: per-event classes, the codegen-bound IDs, and (for TypeScript) the `DestinationKey` union.
```

Author: Addressed in a27de02d. Reframed the doc: single-source (default) vs library mode (alternative for multi-source / shared library setups). Both supported; library is no longer positioned as a successor.

### What the previous output looked like

A single file per source. The `Avo` class held mutable global state, owned the destination SDK, and exposed one method per event:

```swift
let avo = Avo(env: .dev, customDestination: myDestination)
avo.loginSuccess(name: "Alice", clientVersion: "1.2.3")
```

This caused several friction points:

- **Every plan change rewrote the whole file.** Changes to the runtime, the assertion helpers, and the per-event functions all landed in the same diff. Reviewing a property rename meant scrolling past hundreds of lines of unchanged plumbing.
- **One shape of destination wiring per app.** The destination was bound to the `Avo` instance at construction time. Sharing tracking code across apps in a monorepo, or routing the same events to different SDKs in different builds, was awkward.
- **Runtime code couldn't be shared.** Two sources in the same project produced two copies of the runtime — same `AvoAssert`, same `AvoInvoke`, duplicated.

### What's different now

> - **Library file**: the runtime. `AvoEnv`, `AvoAssert`, `AvoInvoke`, the `AvoEvent` protocol/interface, and the `Avo` class. Regenerated rarely.

Reviewer: ```suggestion
- **Library file**: the runtime. `AvoEnv`, `AvoAssert`, `AvoInvoke`, the `AvoEvent` protocol/interface, and the `Avo` class. Regenerated rarely. Shared across all sources/apps. Optionally can be orchestrated from a single place - the "library".
```

Author: Addressed in a27de02d. Reframed the doc: single-source (default) vs library mode (alternative for multi-source / shared library setups). Both supported; library is no longer positioned as a successor.

> - **App file**: per-event types and a thin config object. Regenerated whenever the tracking plan changes.

Reviewer: ```suggestion
- **App file**: per-source types and a thin config object. Regenerated whenever the tracking plan source changes.
```

Author: Addressed in a27de02d. Reframed the doc: single-source (default) vs library mode (alternative for multi-source / shared library setups). Both supported; library is no longer positioned as a successor.

> - **TypeScript only**: a third file, `AvoConfig.ts`, holds the codegen-bound `schemaId`/`actionId`/`branchId`/`sourceId` plus destination API keys. It has no runtime imports — just the values.

Reviewer: I'd suggest to keep it without specific boundaries, say that it's available on some platforms, and then we introduce it when we get a request

```suggestion
- a third file, `AvoConfig.ts`, holds the codegen-bound `schemaId`/`actionId`/`branchId`/`sourceId` plus destination API keys. It has no runtime imports — just the values.
```

Author: Addressed in a27de02d. Reframed the doc: single-source (default) vs library mode (alternative for multi-source / shared library setups). Both supported; library is no longer positioned as a successor.

### Mental model

Events are values you construct. The `Avo` instance turns them into per-destination payloads.

```
LoginSuccessEvent(name = "Alice")  ──┐
                                     ├──> avo.process / avo.track ──> { "Custom": AvoDestinationEvent(...) }
            (a plain data type)      ┘
```

The boundary: **the library file knows nothing about your tracking plan**. The app file knows nothing about how validation or invocation metrics work. The customer decides what to do with the per-destination map the call returns.

> In Swift and Kotlin the returned map is the *only* thing the runtime gives you — fan-out to your analytics SDKs is your job. In TypeScript you can either fan out yourself or pass `AvoDestination` implementations into `Avo.init()` and let the runtime call them.

Reviewer: Skip this sentence, we'll bring the destination interface to Swift/Kotlin too eventually

Author: Removed in a27de02d. Dropped the sentence — destination interface is on the roadmap for Swift/Kotlin too.

### When it isn't the right fit

> - You're on a language other than Kotlin, Swift, or TypeScript. The library approach is implemented only for those three.

Reviewer: Let's not set it as a break-it requirement. We'll implement other platfrorms too when we get a case.

Author: Removed in a27de02d. Dropped the section so platform availability is not framed as a break-it requirement.

- You rely on the previous shape of the `Avo` class (e.g. `avo.loginSuccess(...)` method calls). You'll need to update call sites — see the migration section below.

> - You want the runtime and per-event code in a single file for build-system reasons. The legacy single-file codegen is still supported; don't opt in.

Reviewer: These 2 are good

Author: Thanks — kept those two items in the doc.

---

## 2. Quick start

Pull library-shaped code for an existing source by enabling the library-interface mode on that source, then run `avo pull`. The CLI will emit the library file alongside the existing source file (and `AvoConfig.ts` for TypeScript).

> > **Note:** the exact opt-in flag name surfaced to customers isn't covered here — see Questions / gaps at the end.

Reviewer: They can use the `--forceFeatures` flag for avo pull with corresponding flags: `SwiftLibraryInterface, KotlinLibraryInterface, TypeScriptLibraryInterface`

Author: Fixed in a27de02d. Doc now states `avo pull --forceFeatures SwiftLibraryInterface KotlinLibraryInterface TypeScriptLibraryInterface` for opting in, with a note that workspaces can have it on by default.

### Kotlin

`avo pull` produces:

- `Analytics.kt` — per-event data classes and `AvoTrackingPlanConfig`
- `AnalyticsLibraryInterface.kt` — runtime (`Avo` class, `AvoEvent`, `AvoAssert`, `AvoInvoke`)

Both files share the same package (inferred from the source path, default `sh.avo`).

> **Setup.** Add both files to your Gradle module. Both depend on `org.json` (available on Android out of the box) and `kotlinx.coroutines` if you've enabled coroutines or you're on Kotlin Multiplatform.

Reviewer: Remove this setup paragpaph, kotlinx.coroutines is a different feature flag, not connected to the library interface

Author: Removed in a27de02d. Dropped the kotlinx.coroutines paragraph — separate feature flag, not coupled to library mode.

**Initialize.** Use the `initAvo` extension that's generated alongside the per-event classes:

```kotlin
import sh.avo.Avo
import sh.avo.AvoEnv
import sh.avo.LoginSuccessEvent
import sh.avo.initAvo

val avo = Avo.initAvo(env = AvoEnv.DEV)

> val payloads = avo.process(LoginSuccessEvent(

Reviewer: Avo.process throws AvoVerificationError, let's add a catch similar to the Swift example for reference

Author: Fixed in a27de02d. Added try/catch examples for Kotlin and TypeScript quick starts. Both languages' `process()` / `track()` now show `AvoVerificationError` handling.

    timestamp = 1730000000,
    teamId = "team_42",
    teamDomain = "acme.example"
))

payloads["custom"]?.let { event ->
    myAnalyticsSdk.logEvent(event.name, event.properties)
}
```

`process()` returns `Map<String, AvoDestinationEvent>` keyed by destination name (lowercased identifier from your Avo workspace). The map is empty in noop mode. Validation failures throw `AvoVerificationError` when `strict = true` (the default) and log a warning otherwise.

### Swift

`avo pull` produces:

- `Analytics.swift` — `AvoTypes` namespace, per-event structs, `AvoTrackingPlanConfig`, and an `extension Avo` with a no-config `initAvo`
- `AnalyticsLibraryInterface.swift` — runtime

> The app file does `import Library`. The library file is intended to be packaged as a Swift module named `Library`. See Questions / gaps for how customers are expected to set this up.

Reviewer: I guess it's customer specific, can skip for now

Author: Removed in a27de02d. Skipped the customer-specific section.

### TypeScript

`avo pull` produces three files:

- `Avo.ts` — per-event classes (`LoginSuccessEvent`, etc.) extending `BaseAvoEvent`, plus re-exports of the public runtime surface
- `AvoLibrary.ts` — runtime (`Avo` class, `AvoEvent`, `AvoAssert`, `BaseAvoEvent`)
- `AvoConfig.ts` — codegen-bound config (default export), `DestinationKey` union, and per-destination API keys

`AvoConfig.ts` is the only file that contains your `schemaId`/`actionId`/`branchId`. Move it (and update the import path) to wherever you want `Avo.init` to live — at the app level, a library level, or any module boundary.

**Setup.** Drop the three files into your project. No package manager step beyond what your TypeScript build already does — there are no runtime dependencies outside the standard library and `fetch` (used by `AvoInvoke`).

**Initialize.** `Avo.init` takes two arguments: a runtime `AvoConfig` and the codegen-bound config (default-exported from `AvoConfig.ts`):

```ts
import { Avo, AvoEnv } from './Avo';
import { LoginSuccessEvent } from './Avo';
import codegenConfig from './AvoConfig';

const avo = Avo.init({ env: AvoEnv.Dev }, codegenConfig);

> const payloads = avo.track(new LoginSuccessEvent(1730000000, 'team_42', 'acme.example'));

Reviewer: This one throws too, let's illustrate with a catch

Author: Fixed in a27de02d. Added try/catch examples for Kotlin and TypeScript quick starts. Both languages' `process()` / `track()` now show `AvoVerificationError` handling.

)

// After
val avo = Avo.initAvo(env = AvoEnv.DEV)
val payloads = avo.process(LoginSuccessEvent(
    timestamp = 1730000000,
    teamId = "team_42",
    teamDomain = "acme.example",
))
payloads["custom"]?.let { event ->
    myAnalyticsSdk.logEvent(event.name, event.properties)
}
```

#### Swift

```swift
// Before
let avo = Avo(env: .dev, customDestination: MyCustomDestination())
avo.loginSuccess(
    timestamp: 1730000000,
    teamId: "team_42",
    teamDomain: "acme.example"
)

// After
let avo = Avo.initAvo(env: .dev)
let payloads = try avo.process(event: LoginSuccessEvent(
    timestamp: 1730000000,
    teamId: "team_42",
    teamDomain: "acme.example"
))
if let event = payloads["custom"] {
    MyAnalyticsSDK.logEvent(name: event.name, properties: event.properties)
}
```

#### TypeScript

```ts
// Before
import Avo from './Avo';
Avo.initAvo({ env: Avo.AvoEnv.Dev }, /* destinations */ {});
Avo.loginSuccess({
  timestamp: 1730000000,
  teamId: 'team_42',
  teamDomain: 'acme.example',
});

// After
import { Avo, AvoEnv, LoginSuccessEvent } from './Avo';
import codegenConfig from './AvoConfig';

const avo = Avo.init({ env: AvoEnv.Dev }, codegenConfig);

> avo.track(new LoginSuccessEvent(1730000000, 'team_42', 'acme.example'));

Reviewer: Here we need to either set up the destinations or process manually like in the examples for swift/kt above

Author: Fixed in a27de02d. The Quick start TS example now leads with destinations wired up — same shape as the Swift/Kotlin process() pattern. The manual fan-out variant is documented underneath as the alternative.

Reviewer: Current TS snippets misses the if let event = payloads["custom"] {
    MyAnalyticsSDK.logEvent(name: event.name, properties: event.properties)
} part, isn't it?

Author: Fixed in 54db72a5. Added the per-destination fan-out snippet to the TS "After" example — now matches the Swift/Kotlin coverage.

```

### Checklist of changes in your app

1. **Drop the legacy `Avo` constructor.** Only the legacy `Avo(env:, customDestination:, ...)` constructor (Swift/Kotlin) and the legacy free-function `Avo.initAvo(config, destinations)` (TypeScript) are gone. In the new shape you keep using `Avo.initAvo(...)` for Swift and Kotlin — that's the generated extension on the library `Avo` class — and you switch to `Avo.init(config, codegenConfig)` for TypeScript with the codegen-bound config imported from `AvoConfig.ts`.
2. **Replace event method calls with event constructors and a single `process` / `track` call.** Each `avo.someEvent(...)` becomes `avo.process(SomeEvent(...))` (Swift/Kotlin) or `avo.track(new SomeEvent(...))` (TypeScript).
3. **Move your destination SDK calls to your own code (Swift/Kotlin).** The new `process()` returns a per-destination payload map; the legacy `Avo` class invoked your destinations for you. You're now responsible for calling `myAnalyticsSDK.logEvent(name, properties)`.

> 4. **In TypeScript, choose between manual fan-out and `AvoConfig.destinations`.** If you keep manual fan-out, the migration is closer to Swift/Kotlin. If you wire up `destinations`, you must implement the `AvoDestination` interface (`make`, `logEvent`, `setUserProperties`) for every destination — all-or-nothing.

Reviewer: Highlight that the > you must implement the `AvoDestination` interface option is similar to the old setup. Though the destination type is different, the idea is the same

Author: Fixed in a27de02d. Both the Quick start TS section and the Differences section now state that AvoDestination plays the same role as the single-source custom destination — different shape, same idea.

5. **Update system properties calls (if you use them).** `setSystemProperties(...)` / `Avo.setSystemProperties(...)` becomes `AppSystemProperties.configure(...)` (Kotlin) or `AppSystemProperties.shared.configure({...})` (Swift, TypeScript). Call this once at startup before tracking any events.
6. **Update imports.** Add the library file to your build and import its types where you previously imported from the single `Avo` file. In Swift this is `import Library`; in Kotlin it's `import sh.avo.*` (or whatever package your source path resolves to); in TypeScript it's `import { ... } from './Avo'` (the app file re-exports the public library surface).

### What stays the same

- **Event names, property names, and validation rules.** The tracking plan is unchanged — only the shape of the generated code changed.
- **Strict / noop modes.** `strict` (throw on validation failure in non-prod) and `noop` (drop everything) still work the same way.

> - **Inspector integration.** You still pass your Avo Inspector instance into the init call, and the runtime still forwards events to it.

Reviewer: Also the implementation status works as before.

Author: Fixed in a27de02d. Added implementation status tracking to the What stays the same list.

- **The `avo pull` workflow.** Same command, same authentication, same source/destination configuration in your Avo workspace.
- **Destination API keys.** Still embedded by codegen. In TypeScript they're in `AvoConfig.ts`; in Swift/Kotlin they're held inside `AvoTrackingPlanConfig` and used by the `initAvo` extension.

### Breaking changes to watch for

- **The `Avo` constructor signature.** Both Swift and Kotlin's legacy `Avo` took a `customDestination` argument; the new one doesn't. Old call sites will fail to compile.
- **Per-event method calls.** `avo.someEvent(...)` is gone. The compiler will catch every site that needs to change.

> - **Swift `process(event:)` throws.** The legacy `someEvent(...)` method did not. Wrap in `do / catch` or `try?` depending on whether you want strict-mode validation failures to surface.

Reviewer: Correction - someEvent(...) throws an NSException in old implementation. So this needs to be corrected.

Author: Fixed in a27de02d. Corrected the claim — single-source Swift `someEvent(...)` raised `NSException`. Library mode throws a typed `AvoVerificationError` instead, which the change section now states explicitly.

> - **TypeScript: `track` instead of `Avo.someEvent(...)`.** The free-function shape is replaced by `avo.track(new SomeEvent(...))`. If you previously imported `Avo` and called methods on it, those methods no longer exist on the instance.

Reviewer: Also throws like Swift

Author: Fixed in a27de02d. Added try/catch examples for Kotlin and TypeScript quick starts. Both languages' `process()` / `track()` now show `AvoVerificationError` handling.

- **TypeScript destination shape.** Custom destinations are no longer a single callback object; they implement `AvoDestination` with `make / logEvent / setUserProperties`. Existing custom-destination implementations will not type-check.
- **AppSystemProperties is the only way to set system properties.** The free `setSystemProperties` / static `Avo.setSystemProperties` functions are not generated in library mode. Required system properties not set via `AppSystemProperties.configure(...)` are dropped, with a console warning.

> - **No `setAvoLogger` for Swift.** The library uses `NSLog` directly. Kotlin's `Avo` instance still exposes `setAvoLogger(...)` for redirecting log output.

Reviewer: Something we should fix, not a core limitation to put into the docs

Author: Removed in a27de02d. Agreed — bug to fix, not a doc-worthy limitation.

---

## Questions / gaps

The following could not be verified from the codegen source and need a real answer before this doc ships:

> 1. **CLI opt-in flag.** The codegen distinguishes `#KotlinLibraryInterface`, `#SwiftLibraryInterface`, `#TypeScriptLibraryInterface` feature flags internally (see `AvoCore.res:484–520`). What's the customer-facing way to enable each one — a per-source toggle in the Avo workspace, a `avo.json` field, or a CLI flag on `avo pull`?

Reviewer: Those flags can be enabled with `avo pull --forceFeatures`, see projects/shared/CliFeatureFlag.res

Author: Fixed in a27de02d. Documented the --forceFeatures flags up front in the Quick start section.

> 2. **Swift module setup.** The generated app file does `import Library` (`AvoSwift.res:7457`, default `libraryModuleName = "Library"`). What's the recommended way to create that Swift module — a separate SPM package, an Xcode target, an `internal` framework? Is the library module name configurable per project?

Reviewer: That's just a file, treat it like a file you've wrote yourself. Can use any approach

Author: Fixed in a27de02d. Swift section now says the library file is just a file — package as a Library module, or drop both files in the same module; either works.

> 3. **Kotlin Gradle layout.** Should the library file go in a separate Gradle module or coexist with the app file in the same module? The codegen infers the package from the source path; what happens when customers want different packages for the two files?

Reviewer: Same here - can be anything, just like with the code they've written themselves

Author: Fixed in a27de02d. Kotlin section now says both files are treated like any other Kotlin files — drop them anywhere that fits your module layout.

> 4. **TypeScript: built-in destination wrappers.** The TypeScript app file emits "built-in `AvoDestination` wrapper classes for each plan destination" according to the codegen comments (`AvoTypeScript.res:8875`). I didn't find an example snapshot showing one. Are these emitted when the destination is a known SDK (Amplitude, Mixpanel) and customers just pass them in? Or do customers always write their own `AvoDestination`?

Reviewer: Those are similar to the custom destination, users implement sending events to their analytics destinations by themselves, see platform_tests/typescript/web/LibraryInterfaceApp/src/AvoLibrary.ts

Author: Fixed in a27de02d. TS quick start now leads with the destination implementation (AvoDestination interface) as the recommended starting point. Also added the reference link to platform_tests/typescript/web/LibraryInterfaceApp/src/AvoLibrary.ts.

> 5. **`AppSystemProperties.isConfigured` prerequisite.** The codegen says per-event classes throw if system properties aren't configured (`AvoTypeScript.res:8509–8513`). What's the actual exception type and message customers see? Does it differ between TypeScript, Swift, and Kotlin?

Reviewer: In TS: 

Part of the track method
 let violations: AvoAssertMessage[];
    try {
      violations = event.runValidation();
    } catch (e) {
      if (e instanceof AvoVerificationError) {
        if (this.env !== AvoEnv.Prod && this.strict) {
          throw e;
        } else {
          for (const msg of e.messages) {
            console.error("[avo] Pre-validation check failed: " + msg);
          }
        }
        return payloadMap; // do not fan-out when pre-validation throws, but still return the map
      }
      throw e;
    }

  override runValidation(): AvoAssertMessage[] {
    if (!AppSystemProperties.shared.isConfigured) {
      throw new AvoVerificationError(["AppSystemProperties.configure() must be called before sending events."]);
    }

In Swift:

public func process ...

       let violations = event.collectViolations()
        if !violations.isEmpty {
            if __STRICT__ {
                throw AvoVerificationError(messages: violations.map { $0.getMessage() })
            } else {
                NSLog("[avo] Verification failed: \(violations.map { $0.getMessage() }.joined(separator: ", "))")
            }
        }

    public func verify(fileID: String = #fileID, filePath: String = #filePath, line: Int = #line, column: Int = #column) throws {
        guard AppSystemProperties.shared.isConfigured else {
            throw AvoVerificationError(messages: ["\(filePath):\(line):\(column) [\(fileID)] AppSystemProperties.configure() must be called before sending events."])
        }
        let violations = collectViolations()
        if !violations.isEmpty {
            throw AvoVerificationError(messages: violations.map { "\(filePath):\(line):\(column) [\(fileID)] " + $0.getMessage() })
        }
    }

In Kotlin:

        val violations = event.collectViolations()

        if (__ENV__ != AvoEnv.PROD) {
            AvoInvoke.invoke(event.eventId, event.eventHash, violations, config = config)
        }

        if (violations.isNotEmpty()) {
            if (__STRICT__) {
                throw AvoVerificationError(violations.map { it.message })
            } else {
                Log.w("avo", "[avo] Verification failed: " + violations.joinToString(", ") { it.message })
            }
        }

    override fun verify(): AvoResult<Unit, AvoVerificationError> {
        if (!AppSystemProperties.isConfigured) {
            return AvoResult.Failure(AvoVerificationError(listOf("AppSystemProperties.configure() must be called before sending events.")))
        }
        val violations = collectViolations()
        return if (violations.isEmpty()) {
            AvoResult.Success(Unit)
        } else {
            AvoResult.Failure(AvoVerificationError(violations.map { it.message }))
        }
    }

Author: Fixed in a27de02d. The Differences section now states the AvoVerificationError is what surfaces, with the exact message used when AppSystemProperties.configure() has not been called. Skipped inlining the full verify()/track() snippets to keep the doc focused on customer-facing surface — happy to expand if useful.

> 6. **Inspector setup.** The library exposes a structural `AvoInspector` interface (TypeScript) / `Any?` (Kotlin) / `AnyObject?` (Swift). How do customers obtain an Inspector instance to pass in — is it a separate package, or does the existing Avo Inspector SDK satisfy this interface?

Reviewer: It's same as before, they create an instance of AvoInspector from the SDK and pass it

Author: Fixed in a27de02d. The What stays the same section now explicitly says Inspector setup is unchanged — instantiate from the SDK and pass it into init.

> 7. **Co-existence with the legacy codegen.** Can a single source be flipped from legacy to library mode in place, or must customers create a new source? Is there a deprecation timeline for the legacy single-file output?

Reviewer: Single file is not legacy and will not be deprecated. They serve different use cases. Users can codegen both ways from same source, using the flags

Author: Addressed in a27de02d. Reframed the doc: single-source (default) vs library mode (alternative for multi-source / shared library setups). Both supported; library is no longer positioned as a successor.

> 8. **TypeScript `destinations` partial mode.** The runtime supports "zero destinations" (no fan-out) and "full destinations" (all-or-nothing). What's the recommended pattern for customers who want internal fan-out for some destinations and manual fan-out for others? They appear to have to pick one mode globally.

Reviewer: One mode. full destinations should be starting poiint, if it does not work, switch to "zero destinations"  - that way they can build almost whatever they want

Author: Fixed in a27de02d. TS quick start now leads with the full-destinations pattern as the recommended starting point and shows the zero-destinations / manual fan-out variant underneath as the fallback.

> 9. **Snowplow / Adobe-Experience-Platform / mParticle.** These have their own codegen modules but I didn't verify whether the library approach interacts with them. Is library mode mutually exclusive with destination-specific codegen, or do they compose?

Reviewer: They need to wire those manually, eather through a destination or by using the provided payload from the process/track methods

Author: Fixed in a27de02d. The zero-destinations / manual fan-out path is now documented as the way to wire destinations like Snowplow / Adobe / mParticle directly.

> 10. **Versioning of the library file.** Since the library file is shared runtime, what happens when a customer is on, say, codegen v1.0.0 in one source and v1.1.0 in another that share the same library file? Does `avo pull` regenerate both, or does it warn?

Reviewer: Not at the moment. We'll need to add some mechanism in the future

Author: Fixed in a27de02d. Kept versioning as a single Open items entry noting that no built-in mechanism exists yet for mixed-version sources; regenerate from the same CLI version across sources.
